src/main/java/org/broadinstitute/hellbender/tools/exome/ExomeReadCounts.java

```diff
@@ -0,0 +1,681 @@
+        } else {
+            final List<GenomeLoc> result = new ArrayList<>(intervals.size());
+            for (final Locatable locatable : intervals) {
+                if (locatable == null)
+                   throw new IllegalArgumentException("null interval provided by the engine??? " + Arrays.toString(intervals.toArray()));
+                result.add(genomeLocFactory.createGenomeLoc(locatable.getContig(),locatable.getStart(),locatable.getEnd()));
+            }
+            return result;
+        }
+    }
+
+    private Counts createCounts() {
+        switch(groupBy) {
+            case COHORT: return new TotalCounts();
+            case SAMPLE: return new SampleCounts();
+            case READ_GROUP: return new ReadGroupCounts();
+            default:
+                throw new UserException.BadArgumentValue("groupBy",groupBy.toString(),"unsupported count grouping");
+        }
+
+    }
+
+
+    /**
+     * Opens the output file for writing with a print-writer.
+     * @param output the output file.
+     *
+     * @throws NullPointerException if {@code output} is {@code null}.
+     * @throws UserException.CouldNotCreateOutputFile if there was some problem creating or overwriting {@code output}.
+     * @return never {@code null}.
+     */
+    private PrintWriter openOutputWriter(final File output) {
+        try {
+            return new PrintWriter(output);
+        } catch (final IOException e) {
+            throw new UserException.CouldNotCreateOutputFile(output,e);
+        }
+    }
+
+
+    // Used for outputting progress into the longs
+    private int lastProgressIndex = 0;
+
+    /**
+     * Counts how many base calls a read has aligned on overlapping exons.
+     *
+     * @param read {@inheritDoc}
+     * @param referenceContext {@inheritDoc}
+     */
+    @Override
+    public void apply(final SAMRecord read, final ReferenceContext referenceContext, final FeatureContext featureContext) {
+        final GenomeLoc readLocation = genomeLocFactory.createGenomeLoc(read);
+
+        exonDB.forEachExon(readLocation, (index, exon) -> {
+            counts.count(read, index);
+            // Outputting progress into the logs:
+            if (index > lastProgressIndex && index % 1000 == 0)
+                logger.log(Level.INFO, String.format("Reached exon %d of %d at %s",
+                        index + 1, exonDB.exonCount(), exonDB.exon(lastProgressIndex = index)));
+        });
+    }
+
+    @Override
+    public Object onTraversalDone() {
+        super.onTraversalDone();
+        logger.log(Level.INFO, "...done.");
+        logger.log(Level.INFO,"Writing output file...");
+        writeExonOutputHeaders(outputWriter, summaryPerRowOutputWriter);
+        final int exonCount = exonDB.exonCount();
+        final int countColumnCount = counts.columnCount();
+        final int[] countBuffer = new int[countColumnCount];
```

Comment:
I am done some refactoring that avoid copy all together, so no problem there any longer.